Movie Search/SearchController.swift

```diff
@@ -0,0 +1,122 @@
+        NSObject.cancelPreviousPerformRequests(withTarget: self) 
+        perform(#selector(self.performSearch(_:)),
+                with: movieTitle,
+                afterDelay: Constants.searchThrottleDelay)
+    }
+}
+
+private extension SearchController {
+    
+    class func status(with networkError: NetworkServiceError?) -> SearchStatus? {
+        guard let error = networkError else { return nil }
+        switch error {
+        case .badResponse:
+            return .error("Bad response")
+        case .parsingError:
+            return .error("Parsing error")
+        case .constructingRequestError:
+            return .error("Constructing request error")
+        }
+    }
+    
+    @objc
+    func performSearch(_ movieTitle: String) {
+        if movieTitle.count > 0 {
+            networkService.search(for: movieTitle) { moviesModels, error in
+                self.moviesModels = moviesModels ?? []
+                if let status = Self.status(with: error) {
+                    self.searchPerformedCallback?(status)
+                }
+                else if self.moviesModels.isEmpty {
+                    self.searchPerformedCallback?(.noResult)
+                }
+                else {
+                    self.searchPerformedCallback?(.success)
+                }
+            }
+        }
+        else {
+            moviesModels = []
+            searchPerformedCallback?(.nothingRequested)
+        }
+    }
+    
+    func tryToLoadPhoto(for indexPath: IndexPath, in tableView: UITableView) {
+        let movieModel = moviesModels[indexPath.row]
+        guard movieModel.isPhotoReadyForDownload else { return }
+        movieModel.photoRecord.state = .scheduled
+        
+        let downloadOperation = DownloadOperation(indexPath: indexPath,
+                                                  movieModel: movieModel,
+                                                  networkService: networkService) {
+             tableView.reloadRows(at: [indexPath], with: .fade)
+        }
+        downloadQueue.addOperation(downloadOperation)
+    }
+}
+
+extension SearchController: UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return moviesModels.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: MovieCell.cellIdentifier) as? MovieCell
+        else {
+            fatalError()
+        }
+        
+        let movieModel = moviesModels[indexPath.row]
+        cell.fill(with: movieModel)
+        tryToLoadPhoto(for: indexPath, in: tableView)
+        return cell
+    }
+}
+
+extension SearchController: UITableViewDataSourcePrefetching {
+    
+    func tableView(_ tableView: UITableView, prefetchRowsAt indexPaths: [IndexPath]) {
+        indexPaths.forEach { tryToLoadPhoto(for: $0, in: tableView) }
```

Comment:
so you are prefetching and then fetching again on `cellForRowAt`. There's a good chance you will be duplicating requests if the prefetch hasn't completed and cached the image first.

Reply:
but this method isn't guaranteed to be called https://developer.apple.com/documentation/uikit/uitableviewdatasourceprefetching